libkbfs/disk_block_cache.go

```diff
@@ -163,83 +201,140 @@ func newDiskBlockCacheStandard(config diskBlockCacheConfig, dirPath string,
+		err := cache.config.Codec().Decode(iter.Value(), &metadata)
+		if err != nil {
+			return err
+		}
+		size := uint64(metadata.BlockSize)
+		tlfCounts[metadata.TlfID]++
+		tlfSizes[metadata.TlfID] += size
+		numBlocks++
+		totalSize += size
+	}
+	cache.tlfCounts = tlfCounts
+	cache.numBlocks = numBlocks
+	cache.tlfSizes = tlfSizes
+	cache.currBytes = totalSize
+	return nil
 }
 
 // compactCachesLocked manually forces both the block cache and LRU cache to
 // compact their underlying data.
 func (cache *DiskBlockCacheStandard) compactCachesLocked(ctx context.Context) {
 	cache.blockDb.CompactRange(util.Range{})
-	cache.lruDb.CompactRange(util.Range{})
-	bSize, lruSize, err := cache.getSizesLocked()
-	cache.log.CDebugf(ctx, "Disk Cache bSize=%d lruSize=%d err=%+v", bSize, lruSize, err)
+	cache.metaDb.CompactRange(util.Range{})
+	cache.tlfDb.CompactRange(util.Range{})
 }
 
-// lruKey generates an LRU cache key from a tlf.ID and a binary-encoded block
+// tlfKey generates a TLF cache key from a tlf.ID and a binary-encoded block
 // ID.
-func (*DiskBlockCacheStandard) lruKey(tlfID tlf.ID, blockKey []byte) []byte {
+func (*DiskBlockCacheStandard) tlfKey(tlfID tlf.ID, blockKey []byte) []byte {
 	return append(tlfID.Bytes(), blockKey...)
 }
 
-// updateLRULocked updates the LRU time of a block in the LRU cache to
+// updateMetadataLocked updates the LRU time of a block in the LRU cache to
 // the current time.
-func (cache *DiskBlockCacheStandard) updateLRULocked(ctx context.Context,
-	tlfID tlf.ID, blockKey []byte) error {
-	key := cache.lruKey(tlfID, blockKey)
-	val, err := cache.config.Clock().Now().MarshalBinary()
+func (cache *DiskBlockCacheStandard) updateMetadataLocked(ctx context.Context,
+	tlfID tlf.ID, blockKey []byte, encodeLen int) error {
+	metadata := diskBlockCacheMetadata{
+		TlfID:     tlfID,
+		LRUTime:   cache.config.Clock().Now(),
+		BlockSize: uint32(encodeLen),
+	}
+	encodedMetadata, err := cache.config.Codec().Encode(&metadata)
 	if err != nil {
 		return err
 	}
-	err = cache.lruDb.Put(key, val, nil)
+	err = cache.metaDb.Put(blockKey, encodedMetadata, nil)
 	if err != nil {
 		cache.log.CWarningf(ctx, "Error writing to LRU cache database: %+v", err)
 	}
+	key := cache.tlfKey(tlfID, blockKey)
+	hasKey, err := cache.tlfDb.Has(key, nil)
+	if err != nil {
+		cache.log.CWarningf(ctx, "Error reading from TLF cache database: %+v", err)
+	}
+	if !hasKey {
+		err = cache.tlfDb.Put(key, []byte{}, nil)
+		if err != nil {
+			cache.log.CWarningf(ctx, "Error writing to TLF cache database: %+v", err)
+		}
+	}
 	return nil
 }
 
-// timeFromBytes converts a value from the LRU cache into a time.Time.
-func (*DiskBlockCacheStandard) timeFromBytes(b []byte) (t time.Time, err error) {
-	err = t.UnmarshalBinary(b)
-	return t, err
+// getMetadata retrieves the metadata for a block in the cache, or returns
+// leveldb.ErrNotFound and a zero-valued metadata otherwise.
+func (cache *DiskBlockCacheStandard) getMetadata(blockIDBytes []byte) (
```

Comment:
Actually it's just casting a string as a `[]byte`, so it's probable that it doesn't even `memcpy`.